File: Leetcode/259/259.go
func twoSumLessThanK(nums []int, r int, k int) int {
	i, j := 0, r
	counter := 0
	for i < j {
		temp := nums[i] + nums[j]
		if temp >= k {
			j--
		} else {
			counter += j - i
			i++
		}
	}
	return counter
}

func threeSumSmaller(nums []int, target int) int {
	if len(nums) < 3 {
		return 0
	}
	sort.Ints(nums)
	counter := 0
	for k := 2; k < len(nums); k++ {
        // nums[i] + nums[j] < target - nums[k]
		counter += twoSumLessThanK(nums, k-1, target - nums[k])
	}
	return counter
}

File: Leetcode/259/259.cpp
class Solution {
public:
    int twoSumSmallerThanK(vector<int>& nums, int rightIndex, int target) {
        int l = 0;
        int r = rightIndex;
        int counter = 0;
        int temp;
        while (l < r) {
            temp = nums[l] + nums[r];
            if (temp >= target) {
                --r;
            }
            else {
                counter += r - l;
                ++l;
            }
        }
        return counter;
    }
    // O(n^2)
    int threeSumSmaller(vector<int>& nums, int target) {
        int n = (int)nums.size();
        if (n < 3) {
            return 0;
        }
        sort(nums.begin(),nums.end());
        int counter = 0;
        for (int k = 2; k < n; ++k) {
            counter += twoSumSmallerThanK(nums, k-1, target - nums[k]);
        }
        return counter;
    }
};